lobby/src/main/java/org/triplea/lobby/server/login/forgot/password/verify/TempPasswordVerification.java

```diff
@@ -0,0 +1,36 @@
+package org.triplea.lobby.server.login.forgot.password.verify;
+
+import lombok.AllArgsConstructor;
+import org.mindrot.jbcrypt.BCrypt;
+import org.triplea.lobby.server.db.dao.TempPasswordDao;
+
+/**
+ * Module to verify if a users temporary password is valid, if so then invalidates the temporary
+ * password so that it cannot be used again.
+ */
+@AllArgsConstructor
+public class TempPasswordVerification {
+
+  private final TempPasswordDao tempPasswordDao;
+
+  /**
+   * Validates if a given password matches a temp password for a given user. If the password is
+   * matched, it is invalidated (single-use).
+   *
+   * @param username Username whose password will be checked.
+   * @param hashedPassword Candidate temporary password.
+   * @return True if the hashedPassword matches a temp password for the given user.
+   */
+  public boolean checkTempPassword(final String username, final String hashedPassword) {
+    final String tempPassword = tempPasswordDao.fetchTempPassword(username).orElse(null);
+    if (tempPassword == null) {
+      return false;
+    }
+
+    if (!BCrypt.checkpw(hashedPassword, tempPassword)) {
+      return false;
+    }
+    tempPasswordDao.invalidateTempPasswords(username);
```

Comment:
Maybe that's just me because I always think of password resets in terms of tokens.
But shouldn't the password be invalidated after using it in either case?

Reply:
Slightly confused what the 'either' case refers to. The password here is invalidated if it is used successfully. If we invalidated it on the negative case, then I could try to login as someone else and their temp password would suddenly be invalid. Also if there is a copy/paste mistake and some extra whitespace is captured, a user could invalidate their temp password mistakenly.

I think then this code is correct, let me know  if still see it otherwise.

Reply:
From your perspective the code is indeed correct.
AFAIK common password reset mechanisms send you a link with a token as GET parameter, where it makes perfectly sense to invalidate the actual token when a false token is issued. You possibly can't mistype a hyperlink.
For the password on the other hand it does make sense, as you correctly explained.

Reply:
Indeed! That hyperlink would need to go to a public website with a password reset form. That seemed like a lot more infrastructure and moving parts then issuing a temporary password.

Issuing a temporary password is also not that uncommon, so I don't think we are necessarily doing anything too novel here either.